add feature to convert FP32(float) to FP16(half) using HW instructions

tomoaki0705 opened the pull request:
### What does this PR change?
Allow user to pack 32 bit float data in to 16bit float data (half)
  * check compiler support
  * check HW support before executing
  * add test doing round trip conversion from / to FP32
  * treat array correctly if size is not multiple of 4
  * add declaration to prevent warning

tomoaki0705 commented:
This only packs FP32 to FP16, so it might look not useful.
But thinking of CUDA use, in some case, this could reduce the data to half of the size which would be transferred to GPU.
This might need to be committed on oepncv_contrib repository, but I just pushed it on opencv repository.

It seems that I need to fix some build errors.

tomoaki0705 commented:
OMG, I accidentally pulled and performed a merge on this.
I'm not sure if I can redo that commit, if not, I think I'll just close this pull request and re-do it from the beginning.

vpisarev commented on modules/core/include/opencv2/core/mat.hpp:
```diff
@@ -1084,6 +1084,18 @@ class CV_EXPORTS Mat
     */
     void copyTo( OutputArray m, InputArray mask ) const;
 
+#if CV_FP16
```
1) please, put convertFp16 out of cv::Mat. I'm fine to have it as a separate function.
2) such API cannot be put under conditional compilation; if we add a function to convert to/from fp16, we should always provide it, i.e. there should be software emulation of such function if HW cannot do that

tomoaki0705 replied:
Thank you for the comment.  That's exactly the point I wanted to discuss.

1) I'll follow your suggestion to move convertFp16 as a separate function.  In that case, do you have a suggestion of which module to be added ? Does this still fit in core module ?

2) I understand the background that it'll be better to have a SW emulation, but does this mean that it's a mandatory request to implement the SW emulation version ? let me understand bit more deeper before I start coding the SW emulation.
  a) SW emulation is a must request.  Without that, this PR will be rejected
  b) to have this function in cv::Mat, it's a mandatory to have a SW emulation.  If it's implemented as a separate function out of cv::Mat, it can be under conditional compilation (in another word, comment 1) and 2) are related)
  c) something else

tomoaki0705 replied:
I just moved the function out of cv::Mat.

I figured out that in some cases on ARM 32bit platform, segmentation fault error occurs due to the memory alignment.  So for this, I have no choice but to implement the SW version.  I'll make time and commit a SW version later.

tomoaki0705 commented:
I added SW implementation.
SW result matches HW result according to the test.
Anything else than squashing the commits?

tomoaki0705 commented:
Hello @vpisarev , is there anything else to be done for this PR ?

tomoaki0705 commented:
I rebased AND squashed the commit.

vpisarev commented on cmake/OpenCVCompilerOptions.cmake:
```diff
@@ -146,8 +146,11 @@ if(CMAKE_COMPILER_IS_GNUCXX)
+  endif(ARM)
   if(ENABLE_NEON)
-    add_extra_compiler_option("-mfpu=neon")
+    add_extra_compiler_option("-mfpu=neon-fp16")
```
what's the status of neon-fp16 support? Does any ARMv7 chip support it? What about compiler support? gcc/clang?

tomoaki0705 replied:
- Compiler support
  - I checked and indeed, some early compiler (gcc 4.4) doesn't accept neon-fp16.  I'll put something to guard this part
- Architecture support
  - Technically, fp16 is not architecture support. It's included in FPU. From ARM v8, it's the part of the architecture.
  - This PR includes the check of feature on run time, so It shouldn't be a problem with older processor.

vpisarev commented on modules/core/include/opencv2/core.hpp:
```diff
@@ -524,6 +524,17 @@ For example:
+@param dst output array.
+@param useHW if possible use HW SIMD instruction to convert
+*/
+CV_EXPORTS_W void convertFp16(InputArray src, OutputArray dst, bool useHW = true);
```
useHW parameter should be removed; implementation should determine if the particular instructions are available and use them. It's like we would add parameters like useNEON, useSSE to each single OpenCV function

tomoaki0705 replied:
I agree to your suggestion.
The ONLY reason I left this switch is because of test.
If I remove this, I have to change the test doing round trip between fp16/fp32.
If that's OK for you, I'm happy to remove this switch.

tomoaki0705 commented:
I don't understand the reason why I'm getting such a dozen of test failures on Mac.
The error is happening on EVERY single test, so I guess the cause is from somewhere system.cpp or ts_func.cpp.
But I have no more information about this, and I'd like to ask a help from anyone.

What did I do wrong ?

tomoaki0705 commented:
I tried to fix by removing the fp16 specific modification from ts_func.cpp and system.cpp.
I tried every combination but the result didn't change. This is going to be tough.

mshabunin commented:
Maybe this will give you a hint:
```
Program received signal SIGILL, Illegal instruction.
0x0000000102186228 in cv::initInterTab1D(int, float*, int) ()
   from libopencv_imgproc.3.1.dylib
```
Program crashes during global variable initialization.

tomoaki0705 commented:
Thank you @mshabunin 
That information helps me a lot.
I'll try to dig in.

tomoaki0705 commented:
So I fixed following points
 - option for "neon-fp16"
   - If "neon-fp16" is not available for the compiler, only "neon" will be used
   - If a processor doesn't accept fp16 instruction, it will be checked at runtime
 - useHW option
   - removed
 - Mac error
   - It seems passing mf16c option to gcc, it generates illegual instruction
   - To keep it safe, I placed a guard in cmake

I'm happy to modify if there is any further problem.
Best Regards.

vpisarev commented:
this PR will be recreated because of CI issues
